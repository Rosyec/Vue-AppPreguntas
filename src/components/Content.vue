<script setup lang="ts">
import type { Answer } from '@/interfaces/answer.interface';
import { ref, watch, type Ref, } from 'vue';
import { getAnswer } from "../services/getAnswer.service";

let question: Ref<string> = ref('');
let result: Ref<Answer> = ref({
    answer: '',
    forced: false,
    image: ''
});

watch(question, (value, oldValue) => {
    console.log(value);
    console.log(value);
    if ( value.length === 0 ) result.value = { answer: '', forced: false, image: '' };
    if( !value.includes('?') ) return;
    console.log(value);
    handle.getResult();
});

const getResult = async () => {
    try {
        const answer = await getAnswer();
        result.value = answer;
    } catch (error) {
        console.log('Error Simulado: ', error);
        result.value.answer = 'Fallo la conexion al API';
        result.value.image = '';        
    }
}

const handle = {
    //Esto es para poder testear si la funcion getResult es llamada.
  getResult,
  result
}

defineExpose({ handle });
//Exponemos el objeto handle para poderlo testear.

</script>

<template>
    <div class="container text-center">
        <div>
            <h1>Â¡Realizar Pregunta!</h1><br>
        </div>
        <div class="row">
            <div class="col-md-8">
                <input
                class="form-control"
                type="text"
                name="pregunta"
                placeholder="No olvides dejar al final de cada pregunta un '?'"
                v-model="question">
            </div>
        </div>
        <div class="row">
            <div class="col-md-10">
                <h2 aria-label="answer" class="mt-3" v-show="result.answer"> {{ result.answer === 'yes' ? 'Si' : 'No' }} </h2>
                <img class="mt-3" width="200" height="200" v-show="result.image" :src="result.image" alt="">
            </div>
        </div>
    </div>
</template>

<style scoped>
@font-face {
    font-family: 'Roboto';
    src: url(../assets/fonts/Roboto-Medium.ttf) format("truetype");
}
.container {
    min-height: 90vh;
}

.row {
    justify-content: center;
}

input {
    padding: 0.8em;
    text-align: center;
    border-radius: 999px 999px 999px 999px;
    font-family: Roboto;
    color: brown;
    box-shadow: 20px 20px 100px 0px rgba(0, 0, 0, 0.3);
}

img {
    border-radius: 999px 999px 999px 999px;
}

* {
    margin: 0;
    padding: 0;
}
</style>