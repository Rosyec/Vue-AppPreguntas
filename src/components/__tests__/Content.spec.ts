import createFetchMock from 'vitest-fetch-mock';
import { describe, test, beforeEach, expect, vi } from "vitest";
import 'vitest-fetch-mock';

const fetchMock = createFetchMock(vi);
fetchMock.enableMocks();


import { shallowMount, mount } from "@vue/test-utils";
import Content from "../Content.vue";


describe('Pruebas en <Content/>', () => {
    let wrapper = shallowMount(Content);
    let spyConsole = vi.spyOn( console, 'log' );
    // global.fetch = vi.fn( (): any => Promise.resolve({
    //     json: () => Promise.resolve<Answer>({
    //         answer: 'yes',
    //         forced: false,
    //         image: 'https://yesno.wtf/assets/yes/2.gif'
    //     })
    // }));
    beforeEach( () => {
        wrapper = shallowMount(Content);
        spyConsole = vi.spyOn( console, 'log' );
        vi.clearAllMocks();
        fetchMock.doMock();
    });
    test('Debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });
    test('Escribir en el input no debe de disparar nada, solo el console.log', async () => {
        const spy = vi.spyOn(wrapper.vm.handle, 'getResult');
        const input = wrapper.find('input');
        //Cuando de modifica el DOM se debe poner un await y una funcion async.
        await input.setValue('Hola mundo');
        expect( spyConsole ).toHaveBeenCalledTimes(2);
        expect(spy).toHaveBeenCalledTimes(0);
    });
    test('Disparar el getResult al escribir el simbolo ?', async () => {
        const spy = vi.spyOn(wrapper.vm.handle, 'getResult');
        const input = wrapper.find('input');
        await input.setValue('Hola Mundo?');
        expect( spyConsole ).toHaveBeenCalledTimes(3);
        expect(spy).toHaveBeenCalled();
    });
    test('Pruebas en getAnswer', async () => {
        await wrapper.vm.handle.getResult();
        const { answer, forced, image } = wrapper.vm.handle.result.value;
        const yesno = answer == 'yes' ? 'Si' : 'No';

        const img = wrapper.find('img');
        const h2 = wrapper.find('h2');

        expect( img.attributes('src') ).toBe( image );
        expect( h2.text() ).toBe( yesno );
        
    });
    test('Pruebas en getAnswer - fallo en el Api', async () => {

        fetchMock.mockImplementationOnce( () => Promise.reject('API Error') );

        await wrapper.vm.handle.getResult();

        const { answer, forced, image } = wrapper.vm.handle.result.value;

        expect( answer ).toBe( 'Fallo la conexion al API' );
        expect( image ).toBe( '' );
    });
});


export{}