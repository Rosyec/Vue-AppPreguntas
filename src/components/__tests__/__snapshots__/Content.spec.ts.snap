// Vitest Snapshot v1

exports[`Pruebas en <Content/> > Debe de hacer match con el snapshot 1`] = `
"<div class=\\"container text-center\\" data-v-9ff08d88=\\"\\">
  <div data-v-9ff08d88=\\"\\">
    <h1 data-v-9ff08d88=\\"\\">Â¡Realizar Pregunta!</h1><br data-v-9ff08d88=\\"\\">
  </div>
  <div class=\\"row\\" data-v-9ff08d88=\\"\\">
    <div class=\\"col-md-8\\" data-v-9ff08d88=\\"\\"><input class=\\"form-control\\" type=\\"text\\" name=\\"pregunta\\" placeholder=\\"No olvides dejar al final de cada pregunta un '?'\\" data-v-9ff08d88=\\"\\"></div>
  </div>
  <div class=\\"row\\" data-v-9ff08d88=\\"\\">
    <div class=\\"col-md-10\\" data-v-9ff08d88=\\"\\">
      <h2 aria-label=\\"answer\\" class=\\"mt-3\\" data-v-9ff08d88=\\"\\" style=\\"display: none;\\">No</h2><img class=\\"mt-3\\" width=\\"200\\" height=\\"200\\" src=\\"\\" alt=\\"\\" data-v-9ff08d88=\\"\\" style=\\"display: none;\\">
    </div>
  </div>
</div>"
`;
